packages/reporters/hmr-server/src/HMRServer.js

```diff
@@ -0,0 +1,150 @@
+import https from 'https';
+import WebSocket from 'ws';
+import generateCertificate from '@parcel/server-utils/src/generateCertificate';
+import getCertificate from '@parcel/server-utils/src/getCertificate';
+import logger from '@parcel/logger';
+import prettyError from '@parcel/reporter-cli/src/prettyError';
+
+type HMRAsset = {|
+  id: string,
+  type: string,
+  output: string,
+  deps: Object
+|};
+
+type HMRError = {|
+  message: string,
+  stack?: string
+|};
+
+type HMRMessage = {|
+  type: string,
+  error?: HMRError,
+  assets?: Array<HMRAsset>
+|};
+
+export default class HMRServer {
+  server: Server;
+  wss: WebSocket.Server;
+  unresolvedError: HMRMessage | null = null;
+  changedAssets: Map<string, HMRAsset> = new Map();
+
+  async start(options: ServerOptions) {
+    await new Promise(async resolve => {
+      if (!options.https) {
+        this.server = http.createServer();
+      } else if (options.https === true) {
+        this.server = https.createServer(
+          generateCertificate(options.certificateDir)
+        );
+      } else {
+        this.server = https.createServer(await getCertificate(options.https));
+      }
+
+      let websocketOptions = {
+        server: this.server,
+        origin: options.host
+          ? `${options.https ? 'https' : 'http'}://${options.host}`
+          : undefined
+      };
+
+      this.wss = new WebSocket.Server(websocketOptions);
+      this.server.listen(options.port, resolve);
+    });
+
+    this.wss.on('connection', ws => {
+      ws.onerror = this.handleSocketError;
+
+      if (this.unresolvedError) {
+        ws.send(JSON.stringify(this.unresolvedError));
+      }
+    });
+
+    this.wss.on('error', this.handleSocketError);
+
+    return this.wss._server.address().port;
+  }
+
+  stop() {
+    this.wss.close();
+    this.server.close();
+  }
+
+  emitError(err: PrintableError) {
+    let {message, stack} = prettyError(err);
+
+    // store the most recent error so we can notify new connections
+    // and so we can broadcast when the error is resolved
+    this.unresolvedError = {
+      type: 'error',
+      error: {
+        message,
+        stack
+      }
+    };
+
+    this.broadcast(this.unresolvedError);
+  }
+
+  async addChangedAsset(cacheEntry: CacheEntry) {
+    for (let asset of cacheEntry.assets) {
+      let deps = {};
+      for (let dependency of asset.dependencies) {
+        // ? SourcePath should come from graph?
+        deps[dependency.sourcePath] = dependency.id;
+      }
+
+      this.changedAssets.set(asset.id, {
+        id: asset.id,
+        type: asset.type,
+        output: (await asset.getOutput()).code,
+        deps: deps
+      });
+    }
+  }
+
+  async emitUpdate(reload: boolean = false) {
+    if (this.unresolvedError) {
+      this.unresolvedError = null;
+      this.broadcast({
+        type: 'error-resolved'
+      });
+
+      return;
+    }
+
+    if (reload) {
+      this.broadcast({
+        type: 'reload'
```

Comment:
I think the logic for when to reload the page should be done in loader runtimes. There are a few cases, but mostly it’s based on asset type. HTML should cause a reload, but only if you’re on that page that changed (can’t be determined server side). JS causes a reload unless you called module.hot.accept. CSS always HMR accepts without reloading. If you add a new bundle, page should reload because a script/style tag might have been added to the html page. Probably more too, but the logic for all of that should happen in the loader plugins with custom logic depending on the type of file. Doesn’t have to be in this PR, I’m just writing things down as they come to me.

Reply:
You're probably right already removed a bunch of the reload stuff anyway. Will look for more of that and clean it up